crates/democracy/src/migrations.rs

```diff
@@ -0,0 +1,271 @@
+
+    impl<T: Config + frame_system::Config<Hash = H256>> OnRuntimeUpgrade for Migration<T> {
+        #[cfg(feature = "try-runtime")]
+        fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+            assert_eq!(StorageVersion::get::<Pallet<T>>(), 0, "can only upgrade from version 0");
+
+            let props_count = v0::PublicProps::<T>::get().len();
+            log::info!(target: TARGET, "{} public proposals will be migrated.", props_count,);
+            ensure!(props_count <= T::MaxProposals::get() as usize, "too many proposals");
+
+            let referenda_count = v0::ReferendumInfoOf::<T>::iter().count();
+            log::info!(target: TARGET, "{} referenda will be migrated.", referenda_count);
+
+            Ok((props_count as u32, referenda_count as u32).encode())
+        }
+
+        #[allow(deprecated)]
+        fn on_runtime_upgrade() -> Weight {
+            let mut weight = T::DbWeight::get().reads(1);
+            if StorageVersion::get::<Pallet<T>>() != 0 {
+                log::warn!(
+                    target: TARGET,
+                    "skipping on_runtime_upgrade: executed on wrong storage version.\
+				Expected version 0"
+                );
+                return weight;
+            }
+
+            ReferendumInfoOf::<T>::translate(|index, old: ReferendumInfo<T::BlockNumber, T::Hash, BalanceOf<T>>| {
+                weight.saturating_accrue(T::DbWeight::get().reads_writes(1, 1));
+                log::info!(target: TARGET, "migrating referendum #{:?}", &index);
+                Some(match old {
+                    ReferendumInfo::Ongoing(status) => ReferendumInfo::Ongoing(ReferendumStatus {
+                        end: status.end,
+                        proposal: Bounded::from_legacy_hash(status.proposal),
+                        threshold: status.threshold,
+                        delay: status.delay,
+                        tally: status.tally,
+                    }),
+                    ReferendumInfo::Finished { approved, end } => ReferendumInfo::Finished { approved, end },
+                })
+            });
+
+            let props = v0::PublicProps::<T>::take()
+                .into_iter()
+                .map(|(i, hash, a)| (i, Bounded::from_legacy_hash(hash), a))
+                .collect::<Vec<_>>();
+            let bounded = BoundedVec::<_, T::MaxProposals>::truncate_from(props.clone());
+            PublicProps::<T>::put(bounded);
+            weight.saturating_accrue(T::DbWeight::get().reads_writes(1, 1));
+
+            if props.len() as u32 > T::MaxProposals::get() {
+                log::error!(
+                    target: TARGET,
+                    "truncated {} public proposals to {}; continuing",
+                    props.len(),
+                    T::MaxProposals::get()
+                );
+            }
+
+            // remove old preimages
+            let res = frame_support::migration::clear_storage_prefix(
```

Comment:
Shouldn't we migrate preimages rather than just delete them? (not saying we definitely should, just pointing out the current behavior)

Reply:
AFAIK, Polkadot also just deleted them

Reply:
Yup if we don't mind preimages disappearing then all good

Reply:
Yeah, we don't mind. Just need to time it properly/announce that this will happen.